vcstool/clients/vcs_base.py

```diff
@@ -115,20 +132,145 @@ def test_url(url, retry=2, retry_period=1, timeout=10):
     request.get_method = lambda: 'HEAD'
 
     try:
-        response = urlopen(request)
+        response = _urlopen_retry(retry, retry_period)(urlopen)(
+            request, timeout=timeout)
     except HTTPError as e:
-        if e.code == 503 and retry:
-            time.sleep(retry_period)
-            return test_url(
-                url, retry=retry - 1, retry_period=retry_period,
+        if e.code in (401, 404):
+            # Try again, but with authentication
+            response = _authenticated_urlopen(
+                request, retry=retry, retry_period=retry_period,
                 timeout=timeout)
-        e.msg += ' (%s)' % url
-        raise
+        if response is None:
+            e.msg += ' (%s)' % url
+            raise
     except URLError as e:
-        if isinstance(e.reason, socket.timeout) and retry:
-            time.sleep(retry_period)
-            return test_url(
-                url, retry=retry - 1, retry_period=retry_period,
-                timeout=timeout)
         raise URLError(str(e) + ' (%s)' % url)
     return response
+
+
+def _urlopen_retry(retry, retry_period):
+    def _retry_decorator(f):
+        def _retryable_function(*args, **kwargs):
+            local_retry = retry + 1
+
+            while True:
+                try:
+                    local_retry -= 1
+                    return f(*args, **kwargs)
+                except HTTPError as e:
+                    if e.code != 503 or local_retry <= 0:
+                        raise
+                except URLError as e:
+                    if (not isinstance(e.reason, socket.timeout) or
+                            local_retry <= 0):
+                        raise
+
+                if local_retry > 0:
+                    time.sleep(retry_period)
+                else:
+                    break
+
+        return _retryable_function
+    return _retry_decorator
+
+
+def _authenticated_urlopen(uri, retry=2, retry_period=1, timeout=10):
+    if isinstance(uri, str):
+        request = Request(uri)
+    else:
+        request = uri
```

Comment:
I am wondering if this might be an issue if the passed `uri` is of type `unicode` in Python 2. Maybe the logic should be inverted to:

```
if isinstance(uri, Request):
    request = uri
else:
    request = Request(uri)
```

Reply:
Ah, good catch, this would be `basestring` in python2. I agree, your approach should work better for both. Fixed.